src/main/java/io/cryostat/messaging/MessagingModule.java

```diff
@@ -76,27 +77,22 @@ static MessagingServer provideWebSocketMessagingServer(
             AuthManager authManager,
             NotificationFactory notificationFactory,
             @Named(WS_MAX_CONNECTIONS) int maxConnections,
+            @Named(LIMBO_PRUNER) ScheduledExecutorService limboPruner,
+            Clock clock,
             Logger logger,
-            Gson gson,
-            @Named(WS_WORKER_POOL) ScheduledExecutorService workerPool) {
+            Gson gson) {
         return new MessagingServer(
                 server,
                 env,
                 authManager,
                 notificationFactory,
                 maxConnections,
-                workerPool,
+                limboPruner,
+                clock,
                 logger,
                 gson);
     }
 
-    @Provides
-    @Named(WS_WORKER_POOL)
-    static ScheduledExecutorService provideWorkerPool(
-            @Named(WS_MAX_CONNECTIONS) int maxConnections) {
-        return Executors.newScheduledThreadPool(maxConnections);
```

Comment:
Yes, exactly correct. Very specifically, the old worker pool was needed because each thread from the pool would block waiting for a message from each of the connected clients (up to the maximum), and would handle reading messages from those clients and adding them to a shared queue for incoming command processing. After receiving a message and adding it to the queue the worker would go back to listening for more messages from that client. Another thread was responsible for processing outgoing responses and writing them to all of the clients whenever a message was to be sent out. This resulted in all clients receiving responses corresponding to requests from every client including themselves, so it did actually also act as a notifications channel in many cases.

Now, there is no need for a dedicated listener thread for each client, because we just don't listen for messages from the clients past the first one. That first message is listened to using an async handler so it's done by Vert.x (maybe by a worker pool, maybe by a single thread, not sure - doesn't matter for us anymore) and the message is passed to us. After we verify the subprotocol we remove the message handler and drop whatever messages are sent to us. If the client doesn't send a message after 10s then it's misbehaving and clogging up resources (potentially a Vert.x worker thread, but also 1 "resource" of the maximum number of connected clients we allow), so we drop it, but to do so we need something watching in the background and keeping track of how long the connection has been open without the client authenticating, so a single thread is used for that.